### PR TITLE
expression: refactor builtin function RAND to be compatible with MySQL

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
1. builtinRandSig/builtinRandWithSeedSig.randGen may cause data race when Project executing parallelly.
2. Results of rand(N) is not compatible with MySQL as https://github.com/pingcap/tidb/pull/9387#discussion_r258768167 mentioned.

### What is changed and how it works?
1. Eager evaluate seed for rand() in getFunction.
2. Check whether args[0] is a constant during `getFunction`:
If so, return `builtinRandSig` rather than `builtinRandWithSeedSig`;
If not, evaluate seed each time Rand invoked.

### Check List 

Tests 

integration tests

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/builtin_math.go [hunk 1]
```diff
@@ -1011,12 +1020,16 @@ func (b *builtinRandWithSeedSig) evalReal(row chunk.Row) (float64, bool, error)
...
-			b.randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
-		} else {
-			b.randGen = rand.New(rand.NewSource(seed))
+		b.mu.Lock()
+		if b.randGen == nil {
+			if isNull {
+				// When seed is NULL, it is equal to RAND().
+				b.randGen = rand.New(rand.NewSource(time.Now().UnixNano()))
+			} else {
+				b.randGen = rand.New(rand.NewSource(seed))
```
Comment (lamxTyler) [inline expression/builtin_math.go]: Something not related to this pr, according to mysql, `with a nonconstant initializer argument (such as a column name), the seed is initialized with the value for each invocation of RAND().`, so is it right to only init it once?

Comment (XuHuaiyu) [reply expression/builtin_math.go]: You're right. The implementation of `rand(N)` in TiDB is not compatible with MySQL.

**Comment by XuHuaiyu** [comment]
@zz-jason
Yes, we can, thus we need to add a `Clone` function for `*expression.EvaluatorSuite`. I'll modify it.

File: expression/builtin_math.go [hunk 2]
```diff
@@ -966,8 +966,23 @@ func (c *randFunctionClass) getFunction(ctx sessionctx.Context, args []Expressio
 	bt := bf
 	if len(args) == 0 {
 		sig = &builtinRandSig{bt, nil}
```
Comment (lamxTyler) [inline expression/builtin_math.go]: Can we init the seed here? So we don't need to clone or consider data race.

File: expression/builtin_math.go [hunk 3]
```diff
@@ -966,8 +966,23 @@ func (c *randFunctionClass) getFunction(ctx sessionctx.Context, args []Expressio
...
+		// when the statement is prepared, prior to execution.
+		seed, isNull, err := args[0].EvalInt(ctx, chunk.Row{})
+		if err != nil {
+			return nil, err
+		}
+		if isNull {
+			bf.args = bf.args[1:]
+			seed = time.Now().UnixNano()
+		}
+		bt.args = bt.args[:1]
```
Comment (lamxTyler) [inline expression/builtin_math.go]: Why remove the args?

File: expression/builtin_math.go (hunk 2 above)
Comment (XuHuaiyu) [reply expression/builtin_math.go]: It seems we can.

File: expression/builtin_math.go (hunk 3 above)
Comment (XuHuaiyu) [reply expression/builtin_math.go]: Yep, we do not need to remove it.

**Comment by lamxTyler** [comment]
LGTM

**Comment by zz-jason** [comment]
@XuHuaiyu Please cherry pick this commit to release 2.1 and release 2.0

**Pull request merged**